### PR TITLE
add webserver

### PR DESCRIPTION
**Pull request by taitai42**
ajout d'un webserver permettant d'envoyer des commande depuis le site internet:

eg : une news a ete publie par darwyn : https://www.froggedtv.com/news/8

### CODE REVIEW

File: bot/grenouille_http_server.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                self.end_headers()
+                return
+
+class GrenouilleHttpServer:
+    
+    def __init__(self, grenouille_bot):
+        self.grenouille_bot = grenouille_bot
+        
+    def start(self, grenouille_irc_bot):
+        global bot
```
Comment (snickers54) [inline bot/grenouille_http_server.py]: why a global ?

Comment (taitai42) [reply bot/grenouille_http_server.py]: i don't know how to give it to the do_post function otherwise.

File: bot/grenouille_http_server.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+from http.server import BaseHTTPRequestHandler, HTTPServer
+import threading
+import logging
+import cgi
```
Comment (alex-min) [inline bot/grenouille_http_server.py]: en general tu met les imports systemes en haut et apres les tiens

File: bot/grenouille_http_server.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        global bot
+        bot = grenouille_irc_bot
+
+        try:
+            self.server = threading.Timer(11, self.runserver).start()
+        except Exception as e:
+            logging.info(e)
+
+    def runserver(self):
+        port = 8083
```
Comment (alex-min) [inline bot/grenouille_http_server.py]: tu peux mettre `os.getenv('PORT', 8083)` a la place comme ca on peut changer le port directement avec l'env

Comment (taitai42) [reply bot/grenouille_http_server.py]: c'est toi l'env

Comment (alex-min) [reply bot/grenouille_http_server.py]: ah, pardon alors

File: bot/grenouille_http_server.py (hunk 1 above)
Comment (snickers54) [reply bot/grenouille_http_server.py]: I would use a closure, kind of factory to create your RequestHandler which will have `bot` and you will have access to it

Comment (snickers54) [reply bot/grenouille_http_server.py]: Like 

```
def MakeTaiTaiHandler(bot):
    class testHTTPServer_RequestHandler(BaseHTTPRequestHandler):
        def do_POST():
            print(bot) # this should work
    return testHTTPServer_RequestHandler

myhandler = MakeTaiTaiHandler("test")
myhandler.do_POST() # it's printing "test"
```

Comment (taitai42) [reply bot/grenouille_http_server.py]: cheers, gonna try

**Comment by taitai42** [comment]
- @alex-min  : i'm now using environment var for webserver port. thanks for your review.
- @snickers54 : i removed the global var and used your solution, works fine.
- @Philaeux, @alex-min , @snickers54 : i don't know why but i can't "spam" the command (which is good) i can only use it sometime (i'd say every 60 sec) or if i run another command just after (like !who) i think it's related to the sanitize threading.

I would like to run the webserver from the main class instead of running it in the irc thread, but i don't know how to have the reference to the irc object after, if you guys know, would be nice.

**Comment by Philaeux** [comment]
- A user can't send the same message to a twitch chat too often, this is a IRC server limitation. You can try it on any twitch chat. But moderators don't have this limitation (they can spam as often as they like). So if you are testing in your own private channel, you can set your bot moderator.
- You are missing a secret between the caller and the endpoint to secure your web server. Only a few people should be able to do a post. The best way to do it is to create a GRENOUILLE_API_KEY secret, shared between the bot and the caller to authorize the calls. Pass it in header/parameters. You also have to setup HTTPS to avoid secret stealing.
- The sanitizer is a dirty thing I did because I failed to listen to some Twitch event. I might change it in the future.

I will look into your code in details in the afternoon.

**Comment by Philaeux** [comment]
Sorry https is not necessary, i'll manage the secure endpoint in haproxy. But the api secret is.

**Comment by taitai42** [comment]
oh ok, thanks @Philaeux, didn't think of twitch irc limitations . 

I thought about this stuff, didn't know if it would be that much critical for now, no one is supposed to know the ip address of the bot anyway, and basic people won't know they have to use something like postman to simulate a post request. 

but i'll think about it after work :+1:

File: bot/grenouille_http_server.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+from http.server import BaseHTTPRequestHandler, HTTPServer
+import threading
+import logging
+import cgi
+import os
+
+
+def MakeHandler(bot):
+    class testHTTPServer_RequestHandler(BaseHTTPRequestHandler):
```
Comment (Philaeux) [inline bot/grenouille_http_server.py:9]: Remove `test` here from name.

File: bot/grenouille_http_server.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                if (item.name == "say"):
+                    bot.send_msg(item.value)
+                    
+                    self.send_response(200)
+                    self.send_header("Content-type", "text/html")
+                    self.end_headers()
+                    return
+    return testHTTPServer_RequestHandler
+
+class GrenouilleHttpServer:
```
Comment (Philaeux) [inline bot/grenouille_http_server.py:49]: Would be best if you make this inherit from `Thread` and merge all 3 functions. So you would be able to start it similar to other classes. Timer are not the best way to do it.

Comment (Philaeux) [reply bot/grenouille_http_server.py:49]: Something like:

```
from threading import Thread

class GrenouilleHttpServer(Thread):
   def run(self, grenouille_bot):
      self.bot = grenouille_bot
      port = int(os.environ['WEBSERVER_PORT'])
      server_address = ('127.0.0.1', port)
      myhandler = MakeHandler(self.bot)
      httpd = HTTPServer(server_address, myhandler)
      logging.info('webserver started on port ' + str(port))
      httpd.serve_forever()
```

After you add `self.grenouille_http_server.start()` in the main class. This way you don't have to cross thread starting by calling it in `grenouille_irc_bot.py`

File: bot/grenouille_irc_bot.py [hunk 6]
```diff
@@ -62,6 +62,8 @@ def on_welcome(self, connection, e):
         connection.send_raw('CAP REQ :twitch.tv/tags')
         logging.info('Connected to channel.')
 
+        self.grenouille_bot.grenouille_http_server.start(self)
```
Comment (Philaeux) [inline bot/grenouille_irc_bot.py]: You can remove this

File: docker/docker-compose.yml [hunk 7]
```diff
@@ -8,3 +11,9 @@ services:
       context: ..
       dockerfile: docker/grenouillebot/Dockerfile
     entrypoint: python3 /grenouillebot/bot/bot_application.py
+    environment:
+        WEBSERVER_PORT: 8083
+    ports:
+      - "8083:8083"
+    network_mode: "host"
+
```
Comment (Philaeux) [inline docker/docker-compose.yml]: Add newline at end of file.

File: bot/grenouille_irc_bot.py [hunk 8]
```diff
@@ -79,6 +81,9 @@ def on_ping(self, connection, e):
         """
         self.last_ping = datetime.utcnow()
 
+    def send_msg(self, line):
+        self.connection.privmsg(self.channel, line)
```
Comment (Philaeux) [inline bot/grenouille_irc_bot.py]: We will have to secure the `send_msg` somehow. If we send the message when the bot is offline (sometimes), the process will crash. That's just a sidenote for you, I will do a small patch to improve downtime behavior.

**Comment by taitai42** [comment]
hey @Philaeux 
- I added the secret key stuff
- I refactored grenouille_http class to start it from the bot_application class and not after irc is connected
- I secured the send_msg by adding a try catch, it's light for now, a better solution would be to push message in an object and send them once we are connected again, but that's not the purpose of this pull request.

tell what you think about this right now  ?

**Comment by Philaeux** [comment]
Seems to work quite fine. I will do some rework to turn it into a thread, comment, and I will merge it. After deployment, the grenouillebot webserver will be accessible at `grenouillebot.the-cluster.org`. I will send you a generated secret by Skype

**Comment by snickers54** [comment]
@Philaeux a lot of github accounts have been hacked, I wouldn't consider Skype as a secure channel to give such a secret.

**Comment by taitai42** [comment]
@snickers54  it's fine we have our channel for the froggedtv project, with trusted accounts. 

@Philaeux it's already running in a thread, not sure I understood your comment. let me know when you deploy this so I can deploy the code in the froggedtv website too

**Pull request merged**